### ISSUE TITLE
Check header patch

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#2314]
utils/confgettext.py:
- change fixed header variable to one including proper date and alike

utils/buildcat.py:
- import detect_revision to have a single implementation
- modify options for xgettext to include header tags
- Add header check
   This is dirty! Beware. I intended to create a function that would be as careful as I can to the original source, but still check the most common (to me) errors and update the svn revision automagically.

Plot:

Create array of regular expressions used.
Create array of templates for output.
Match up to line 16 of the source file against regex's.
- Return if empty line encountered -> headers end here
- If line twelve, use special case rule to check list address.
- On all other lines, use regex array and if no match, print informative message.

The function always returns an array, which may be used to choose actions on the lines just checked. I only use one (hopefully) minimal intrusive change method to line 7, proposed to contain the svn revision. It changes the revision number silently, but only after the warning was already produced. This is suboptimal at least.

I tried with the current de_DE files (utils/buildcat.py de_DE). They produce some warings, on second run these fade to one warning for each file. I decided to use full .po-file paths in the first headerline for the script after commiting yesterday.

Comments are very welcome. I would like to upload this or similar to svn soon. Do you think this should be done?

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2007-04-24T17:36:00Z
 * owner: azagtoth
 * the launchpad url was https://bugs.launchpad.net/bugs/535887

**Comment by bunnybot** [comment]
(by atrueba-users)
Logged In: YES 
user_id=544332
Originator: NO

Comments about this:

**Comment by bunnybot** [comment]
(by atrueba-users)
Logged In: YES 
user_id=544332
Originator: NO

(Forget previous comment - wrong key was pressed O:-) )

- You force copyright line to the range of years 2005-200x, whereas it should start in the year of the first edition of the file (there was no Spanish translation in 2005, so no file copyright either).

- Lines #1 and #7 (also, see comment below) can be silently updated, no need to warn the user about mismatchs.

- "Language-Team" check issued a warning in my files ("Language-Team: Spanish/Spain <email address hidden>\n") because of language not being a single word. Strictly speaking, ISO-639 language codes of the form ll_LL describe a pair of language/Area, so should also be allowed.

- I got a warning about invalid additional headers (which aren't there), but it didn't told me the line number.

- You rely on lines being in a fixed place and order but, what if the user adds more comments? or if his program stores headers in a different order? A better approach would be to store all header lines (up to the first blank line) in an array, and check every line of the array against all "updatable" regexps, update those that match, and ignore those that don't.

**Comment by bunnybot** [comment]
(by azagtoth)
Logged In: YES 
user_id=1725302
Originator: YES

I reworked the do_header_check function and its return handling.

The result is cleaner, and way more intrusive.

You can disable the rewrite and check altogether by setting the environment variable NO_HEADER_REWRITE. Please if someone could do this via parameter parsing, I'd be glad.

If this is not set, the function will mercilessly rewrite headers. I ran buildcat with
options de_DE, es_ES and -a. I can't just now tell entirely for certain,
but random checks make clear that at least some files survived this.

Do you think this will do? If one changes the regular expressions in re_compiled_array,
it is easily controlled which lines will get rewritten at all.

And of course it would be nice to have some more candy. But not tonight...

BTW: no changes to the confgettext.py patch.
File Added: utils-gettext.diff.2

**Comment by bunnybot** [comment]
(by atrueba-users)
Logged In: YES 
user_id=544332
Originator: NO

Way cleaner output now ;-) 

The only point here is that you assume that "Plural-Forms" header always has a constant content, while it is user-definable: in my case, the script drops the correct line "Plural-Forms: nplurals=INTEGER; plural=EXPRESSION;\n" (no, this header is not in SVN. I'll commit this updated catalogs anytime soon).

Just update the regexp to something like "re.compile(r"^\"Plural-Forms:")" and you're done.

**Comment by bunnybot** [comment]
(by azagtoth)
Logged In: YES 
user_id=1725302
Originator: YES

Included slightly modified second version in source.
